src/libraries/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs

```diff
@@ -56,32 +57,120 @@ public static unsafe EventWaitHandle Create(bool initialState, EventResetMode mo
+        public static EventWaitHandle OpenExisting(string name, EventWaitHandleRights rights)
         {
-            int errorCode = Marshal.GetLastWin32Error();
+            switch (OpenExistingWorker(name, rights, out EventWaitHandle? result))
+            {
+                case OpenExistingResult.NameNotFound:
+                    throw new WaitHandleCannotBeOpenedException();
+
+                case OpenExistingResult.NameInvalid:
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
 
-            if (handle.IsInvalid)
+                case OpenExistingResult.PathNotFound:
+                    throw new IOException(SR.Format(SR.IO_PathNotFound_Path, name));
+
+                case OpenExistingResult.Success:
+                default:
+                    Debug.Assert(result != null, "result should be non-null on success");
+                    return result;
+            }
+        }
+
+        /// <summary>
+        /// Tries to open a specified named event wait handle, if it already exists, applying the desired access rights, and returns a value that indicates whether the operation succeeded.
+        /// </summary>
+        /// <param name="name">The name of the event wait handle to be opened. If it's prefixed by "Global", it refers to a machine-wide event wait handle. If it's prefixed by "Local", or doesn't have a prefix, it refers to a session-wide event wait handle. Both prefix and name are case-sensitive.</param>
+        /// <param name="rights">The desired access rights to apply to the returned event wait handle.</param>
+        /// <param name="result">When this method returns, contains an object that represents the named event wait handle if the call succeeded, or <see langword="null" /> if the call failed. This parameter is treated as uninitialized.</param>
+        /// <returns><see langword="true" /> if the named event wait handle was opened successfully; otherwise, <see langword="false" />.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="name"/> is <see langword="null" /></exception>
+        /// <exception cref="ArgumentException"><paramref name="name"/> is an empty string.</exception>
+        /// <exception cref="IOException">A Win32 error occurred.</exception>
+        /// <exception cref="UnauthorizedAccessException">The named event wait handle exists, but the user does not have the security access required to use it.</exception>
+        public static bool TryOpenExisting(string name, EventWaitHandleRights rights, out EventWaitHandle? result) =>
+            OpenExistingWorker(name, rights, out result!) == OpenExistingResult.Success;
+
+        private static OpenExistingResult OpenExistingWorker(string name, EventWaitHandleRights rights, out EventWaitHandle? result)
```

Comment:
Nullability question: The `out EventWaitHandle? result` parameter is not null when the return value is the enum value `OpenExistingResult.Success`. How do I express that as an attribute for the parameter?

The following is not possible because `returnValue` is boolean:
```cs
[NotNullWhen(returnValue: OpenExistingResult.Success)]
```

Reply:
Doesn't look like there's a way to express that. Could either override it in the caller to satisfy its nullability if necessary or change this method to return `bool` and take the result in another out parameter.

Reply:
I requested it in dotnet/roslyn: https://github.com/dotnet/roslyn/issues/48463